### ISSUE TITLE
Filter secrets from Pulumi's outputs

### ISSUE THREAD

**Issue by joeduffy** [pulumi/pulumi#747]
At the moment, we try to `***` secrets by default.  And, in the future, we will have [a `Secret<T>` type](https://github.com/pulumi/pulumi/issues/397), which will minimize risk of disclosure.

But the reality is, there are plenty of circumstances when secrets will get emitted.  For instance, many apps stick them into container environment variables, which will be printed during previews.

I think we can steal a page from Travis here.  See https://blog.travis-ci.com/2017-05-08-security-advisory; in particular:

> With the second issue, we needed to implement a fix that would operate at runtime within the build, regardless of its components. This was solved initially by updating travis-build to automatically filter secure environment variables from the logs for any new build. This is currently deployed in production on both .org and .com. If you are an Enterprise customer you’ll receive this fix by May 12th. We will notify you with information about how to update your installation.

In short, all we need to do is search-and-replace known secrets in the output, and `***` them out.

/cc @CyrusNajmabadi who initially had this idea
